### ISSUE TITLE
Lint idea: upcast function literals by parameter type

### ISSUE THREAD

**Issue by MichaelRFairhurst** [dart-lang/sdk#58017]
Related to https://github.com/dart-lang/sdk/issues/25490

Potential name ideas: `inexact_function_literal_types`.

In some cases, for instance `fold`, we're able to infer the parameter type to the invocation but not the arguments:

```dart
void f(List<Person> people) {
  final byName = people.fold({}, (map, next) {
    // parameter "map" is of type "dynamic" here
   ...
  }); // but fold is inferred as fold<Map>!
}
```

This is spec behavior but not ideal behavior.

For better type safety, it is a good idea to explicitly write `fold<Map>(...)` because then downward inference gets the correct parameter type in the function literal. Or you can write `(Map map, ...` in the function literal.

This is kind of like no-implicit-dynamic, however, this problem can occur with any type not just dynamic (if type signature of fold were `R fold<R extends Object>(...` we'd get an implicit `Object` not an implicit `dynamic`.).

And it could tackle the problem differently, by recognizing all unuseful function casts:

```
ints.fold<int>(0, (num x, num y) => ...); // "num" parameters are less specific than they could be!
```

It is *not* without potential false postiives though, as in theory, the implementation of `fold` could cast the fold function to an arbitrary type, and then having a runtime type that's wider than the static parameter type could be important to get a program to run correctly.

@leafpetersen @jmesserly for thoughts.

**Comment by srawlins** [comment]
Agreed; I think `--strict-inference` is just the way we'll support this.
